File: 1624-clone-binary-tree-with-random-pointer/clone-binary-tree-with-random-pointer.ts
/**
 * Definition for _Node.
 * class _Node {
 *     val: number
 *     left: _Node | null
 *     right: _Node | null
 *     random: _Node | null
 * 
 *     constructor(val?: number, left?: _Node | null, right?: _Node | null, random?: _Node | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *         this.random = (random===undefined ? null : random)
 *     }
 * }
 */


function copyRandomBinaryTree(root: _Node | null): NodeCopy | null {
    if (!root) return null;

    let queue: Node[] = [root];
    let mapNodeCopy = new Map<Node, NodeCopy>();
    mapNodeCopy.set(root, new NodeCopy(root.val));

    while (queue.length > 0) {
        const node: Node = queue.shift();
        // since the newNodeCopy is a reference to an object,
        // all data that changes in one iteration will change
        // the object in the hash map
        const newNodeCopy: NodeCopy = mapNodeCopy.get(node);
        
        if (node.left) {
            if (!mapNodeCopy.get(node.left)) {
                mapNodeCopy.set(node.left, new NodeCopy(node.left.val));
                queue.push(node.left);
            }
            newNodeCopy.left = mapNodeCopy.get(node.left);
        }

        if (node.right) {
            if (!mapNodeCopy.get(node.right)) {
                mapNodeCopy.set(node.right, new NodeCopy(node.right.val));
                queue.push(node.right);
            }
            newNodeCopy.right = mapNodeCopy.get(node.right);
        }

        if (node.random) {
            if (!mapNodeCopy.get(node.random)) {
                mapNodeCopy.set(node.random, new NodeCopy(node.random.val));
                queue.push(node.random);
            }
            newNodeCopy.random = mapNodeCopy.get(node.random);
        }
    }
    
    return mapNodeCopy.get(root);
};

File: 1624-clone-binary-tree-with-random-pointer/clone-binary-tree-with-random-pointer.py
# Definition for Node.
# class Node:
#     def __init__(self, val=0, left=None, right=None, random=None):
#         self.val = val
#         self.left = left
#         self.right = right
#         self.random = random

class Solution:
    def copyRandomBinaryTree(self, root: 'Optional[Node]') -> 'Optional[NodeCopy]':
        if not root:
            return None

        old_to_new = {}

        def dfs(node):
            if not node:
                return None

            if node in old_to_new:
                return old_to_new[node]

            copy = NodeCopy(node.val)
            old_to_new[node] = copy

            copy.left = dfs(node.left)
            copy.right = dfs(node.right)
            copy.random = dfs(node.random)

            return copy

        return dfs(root)